### PR TITLE
Refactor onDomRefresh

### PR DESCRIPTION
**Pull request by jmeas**
WIP

Todo:

- [ ] Add tests back in
- [ ] Refactor onDomRefresh

ref https://github.com/marionettejs/backbone.marionette/issues/1887

### CODE REVIEW

**Comment by thejameskyle** [comment]
:tada:

**Comment by jasonLaster** [comment]
Nice. Let's make sure this is deprecated on minor

File: src/abstract-view.js [hunk 1]
```diff
@@ -25,7 +27,10 @@ Marionette.AbstractView = Backbone.View.extend({
 
     Backbone.View.apply(this, arguments);
 
-    Marionette.MonitorDOMRefresh(this);
+    this.on({
+      render: this._setRendered,
+      show: this._setShown
+    }, this);
```
Comment (jmeas) [inline src/abstract-view.js]: I know, I know. This is odd. The thing is, `onDomRefresh` also set internal state of Views that the CollectionView relies on. We need to continue tracking that internal state for the views to continue functioning in the same way.

**Comment by jmeas** [comment]
Updated!

**Comment by jmeas** [comment]
Updated @paulfalgout! Good catch.

File: docs/marionette.abstractview.md [hunk 2]
```diff
@@ -17,7 +17,6 @@ AbstractView directly.
 * [AbstractView onBeforeDestroy](#abstractview-onbeforedestroy)
 * [AbstractView "attach" / onAttach event](#abstractview-attach--onattach-event)
 * [AbstractView "before:attach" / onBeforeAttach event](#abstractview-beforeattach--onbeforeattach-event)
-* [AbstractView "dom:refresh" / onDomRefresh event](#abstractview-domrefresh--ondomrefresh-event)
```
Comment (samccone) [inline docs/marionette.abstractview.md]: not in the .yaml?

File: src/abstract-view.js [hunk 3]
```diff
@@ -9,6 +9,8 @@
 
 Marionette.AbstractView = Backbone.View.extend({
   isDestroyed: false,
+  _isShown: false,
```
Comment (samccone) [inline src/abstract-view.js]: who uses these...

Comment (jmeas) [inline src/abstract-view.js]: CollectionView. [Comment here](https://github.com/marionettejs/backbone.marionette/pull/2281#discussion-diff-24139170)

Comment (samccone) [inline src/abstract-view.js]: ah cra

Comment (paulfalgout) [inline src/abstract-view.js]: `CompositeView` uses _isRendered  
There's really not much reason these couldn't move into those two.
`CollectionView` has that '_onShowCalled` method and `CompositeView` has a `render` function.  I don't think there'd be need for the listeners.

Comment (jmeas) [inline src/abstract-view.js]: :+1: @paulfalgout.

Comment (samccone) [inline src/abstract-view.js]: yep yep @paulfalgout want to open a new issue for this?

Comment (paulfalgout) [inline src/abstract-view.js]: https://github.com/marionettejs/backbone.marionette/issues/2287

**Comment by jmeas** [comment]
Rebased – let's get this in!

**Comment by lcwork** [comment]
I need to use `onDomRefresh` in a few places in my app. `onAttach` does not work in these cases. Will there be an alternate way to handle a refresh of an existing view by 3.0? (Or one that exists already?) 

I did see a comment about `onRefresh`, but only saw it mentioned in one place. Thanks.

**Comment by jmeas** [comment]
What's the use case?

**Comment by samccone** [comment]
wI could invent one... every render of a view you need to calc the bounding rect of the view to do some jquery plugin goop. (epic editor for instance does this) 
on attach would only work once and not on each rerender, but perhaps you could just use onRender after onAttach fired once... just some thoughts :)

**Comment by lcwork** [comment]
Saving something to the server and the server is massaging the data a bit, and we want that to be shown the user. After I get the data back I can show the new data by rendering, but I need to call a plugin (X-Editable) to make the data "editable" again. I tried a bunch of different things and onDomRefresh was the the only thing that worked.

**Comment by paulfalgout** [comment]
@lcwork We played with a similar situation and was able to reproduce that functionality by doing what @samccone mentioned.   Essentially:

```
onAttach: function(){
  this.doSomething();
  this.on('render', this.doSomething);
}
```

**Comment by lcwork** [comment]
@paulfalgout okay thanks. I will try that.

**Comment by jmeas** [comment]
I could see us adding that back into the core library, if people were interested. Maybe the name `onDomRefresh` could stick around even. But it can be added in fewer LoC, and be less complicated.

Also, it should work with the view being detached from a region, then reattached – i have a feeling the current implementation does not.

**Comment by samccone** [comment]
:+1: yep @paulfalgout that is what i was thinking

**Comment by jmeas** [comment]
@lcwork see #2299. Thanks for the comments :)

File: docs/marionette.abstractview.md (hunk 2 above)
Comment (jmeas) [inline docs/marionette.abstractview.md]: Nope. It's just a stub with not much content.

**Comment by jmeas** [comment]
Ready for review

**Comment by samccone** [comment]
@jmeas I think your commit message description is messed up ha

other than that :shipit:

**Comment by jmeas** [comment]
lol vimfail. Updated.

Should I add a new `onDomRefresh` and leave the tests in this same PR?

**Comment by samccone** [comment]
I would add it in a new PR since it is a "new" feature, what do you think?

**Comment by jmeas** [comment]
Fine by me!

**Comment by samccone** [comment]
cool, one more set of eyes and we can get this in!

**Comment by jasonLaster** [comment]
I think it'd be better to do the refactor in one PR. That way, the refactor is *atomic*

**Comment by samccone** [comment]
kk

**Comment by samccone** [comment]
ping ping @jmeas ok looks like we still need the following

* [ ] reintroduce refactored onDomRefresh
* [ ] put tests back in place around this behavior
* [ ] add a onDomRefresh test that would have failed before and now wont :speak_no_evil: 
* [ ] add back in documentation


Let me know if you want a hand doing this @jmeas thanks again for this amazing :construction_worker:

**Comment by jasonLaster** [comment]
@jmeas mind if I close this and we can return to it when there's interest in picking it up again?

We'll still have the open issue

**Comment by jasonLaster** [comment]
Closing as it has gone stale, but it's referenced in open issues so can be picked up later

**Pull request closed without merging**